File: backend/src/utils/pdfProcessor.js
import fs from 'fs';
import path from 'path';

// Extract text from PDF buffer (simplified version)
// In production, you'd use a library like pdf-parse or pdf2pic
export const extractTextFromPDF = async (pdfBuffer) => {
  try {
    // For now, we'll simulate PDF text extraction
    // In a real implementation, you'd use pdf-parse:
    // const pdfParse = require('pdf-parse');
    // const data = await pdfParse(pdfBuffer);
    // return data.text;
    
    // Simulated extraction - return placeholder text
    return `
    Extracted PDF Content:
    
    This is a simulated PDF text extraction. In a production environment, 
    this would contain the actual text content from the uploaded PDF file.
    
    The PDF processing system would:
    1. Parse the PDF structure
    2. Extract text content from all pages
    3. Preserve formatting where possible
    4. Handle images and tables appropriately
    5. Return clean, analyzable text
    
    Sample research paper content would appear here, including:
    - Abstract sections
    - Introduction and methodology
    - Results and discussion
    - References and citations
    - Statistical data and figures
    
    This extracted text would then be passed to the verification service
    for comprehensive authenticity analysis.
    `;
    
  } catch (error) {
    console.error('Error extracting text from PDF:', error);
    throw new Error('Failed to extract text from PDF file');
  }
};

// Validate PDF file
export const validatePDFFile = (file) => {
  const errors = [];
  
  // Check file size (max 10MB)
  const maxSize = 10 * 1024 * 1024; // 10MB
  if (file.size > maxSize) {
    errors.push('PDF file size must be less than 10MB');
  }
  
  // Check file type
  const allowedTypes = ['application/pdf'];
  if (!allowedTypes.includes(file.mimetype)) {
    errors.push('Only PDF files are allowed');
  }
  
  // Check file extension
  const allowedExtensions = ['.pdf'];
  const fileExtension = path.extname(file.originalname).toLowerCase();
  if (!allowedExtensions.includes(fileExtension)) {
    errors.push('File must have .pdf extension');
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

// Process uploaded PDF for verification
export const processPDFForVerification = async (fileBuffer, fileName) => {
  try {
    // Extract text from PDF
    const extractedText = await extractTextFromPDF(fileBuffer);
    
    // Basic text validation
    if (!extractedText || extractedText.trim().length < 100) {
      throw new Error('PDF appears to be empty or contains insufficient text for analysis');
    }
    
    // Clean and prepare text for analysis
    const cleanedText = cleanTextForAnalysis(extractedText);
    
    return {
      success: true,
      extractedText: cleanedText,
      metadata: {
        fileName,
        textLength: cleanedText.length,
        wordCount: cleanedText.split(/\s+/).filter(w => w.length > 0).length,
        extractedAt: new Date().toISOString()
      }
    };
    
  } catch (error) {
    console.error('Error processing PDF:', error);
    return {
      success: false,
      error: error.message || 'Failed to process PDF file'
    };
  }
};

// Clean extracted text for analysis
const cleanTextForAnalysis = (text) => {
  return text
    // Remove excessive whitespace
    .replace(/\s+/g, ' ')
    // Remove page numbers and headers/footers patterns
    .replace(/page\s+\d+/gi, '')
    .replace(/^\d+\s*$/gm, '')
    // Remove common PDF artifacts
    .replace(/\f/g, ' ') // Form feed characters
    .replace(/[\x00-\x1F\x7F]/g, '') // Control characters
    // Normalize line breaks
    .replace(/\r\n/g, '\n')
    .replace(/\r/g, '\n')
    // Clean up multiple newlines
    .replace(/\n{3,}/g, '\n\n')
    .trim();
};

// Extract metadata from PDF (placeholder)
export const extractPDFMetadata = async (pdfBuffer) => {
  try {
    // In production, you'd extract real PDF metadata
    // using libraries like pdf-lib or pdf-parse
    
    return {
      title: 'Extracted PDF Title',
      author: 'PDF Author',
      subject: 'PDF Subject',
      creator: 'PDF Creator',
      producer: 'PDF Producer',
      creationDate: new Date().toISOString(),
      modificationDate: new Date().toISOString(),
      pageCount: 1,
      fileSize: pdfBuffer.length
    };
    
  } catch (error) {
    console.error('Error extracting PDF metadata:', error);
    return null;
  }
};

// Check if PDF is password protected
export const isPDFPasswordProtected = (pdfBuffer) => {
  try {
    // Simple check for encrypted PDF markers
    const pdfString = pdfBuffer.toString('binary');
    return pdfString.includes('/Encrypt') || pdfString.includes('/Filter/Standard');
  } catch (error) {
    return false;
  }
};

// Estimate PDF complexity for processing time
export const estimatePDFComplexity = (pdfBuffer) => {
  try {
    const sizeInMB = pdfBuffer.length / (1024 * 1024);
    
    if (sizeInMB < 1) return 'low';
    if (sizeInMB < 5) return 'medium';
    return 'high';
    
  } catch (error) {
    return 'unknown';
  }
};

File: backend_python/services/pdf_processor.py
import PyPDF2
import logging
from io import BytesIO
from typing import Optional
import re

logger = logging.getLogger(__name__)

class PDFProcessorService:
    def __init__(self):
        pass

    async def extract_text_from_pdf(self, pdf_content: bytes) -> str:
        """Extract text from PDF content"""
        try:
            logger.info("Extracting text from PDF")
            
            # Create BytesIO object from PDF content
            pdf_stream = BytesIO(pdf_content)
            
            # Create PDF reader
            pdf_reader = PyPDF2.PdfReader(pdf_stream)
            
            # Check if PDF is encrypted
            if pdf_reader.is_encrypted:
                raise Exception("PDF is password protected and cannot be processed")
            
            # Extract text from all pages
            full_text = ""
            for page_num in range(len(pdf_reader.pages)):
                try:
                    page = pdf_reader.pages[page_num]
                    page_text = page.extract_text()
                    full_text += page_text + "\n"
                except Exception as e:
                    logger.warning(f"Error extracting text from page {page_num + 1}: {str(e)}")
                    continue
            
            # Clean the extracted text
            cleaned_text = self._clean_extracted_text(full_text)
            
            if len(cleaned_text.strip()) < 50:
                raise Exception("PDF contains insufficient readable text")
            
            logger.info(f"Successfully extracted {len(cleaned_text)} characters from PDF")
            return cleaned_text
            
        except Exception as e:
            logger.error(f"Error extracting text from PDF: {str(e)}")
            raise Exception(f"Failed to extract text from PDF: {str(e)}")

    def _clean_extracted_text(self, text: str) -> str:
        """Clean extracted text for analysis"""
        if not text:
            return ""
        
        # Remove excessive whitespace
        text = re.sub(r'\s+', ' ', text)
        
        # Remove page numbers and headers/footers patterns
        text = re.sub(r'page\s+\d+', '', text, flags=re.IGNORECASE)
        text = re.sub(r'^\d+\s*$', '', text, flags=re.MULTILINE)
        
        # Remove common PDF artifacts
        text = text.replace('\f', ' ')  # Form feed characters
        text = re.sub(r'[\x00-\x1F\x7F]', '', text)  # Control characters
        
        # Normalize line breaks
        text = text.replace('\r\n', '\n')
        text = text.replace('\r', '\n')
        
        # Clean up multiple newlines
        text = re.sub(r'\n{3,}', '\n\n', text)
        
        # Remove leading/trailing whitespace
        text = text.strip()
        
        return text

    def validate_pdf_file(self, file_size: int, content_type: str, filename: str) -> dict:
        """Validate PDF file before processing"""
        errors = []
        
        # Check file size (max 10MB)
        max_size = 10 * 1024 * 1024  # 10MB
        if file_size > max_size:
            errors.append('PDF file size must be less than 10MB')
        
        # Check content type
        if content_type != 'application/pdf':
            errors.append('Only PDF files are allowed')
        
        # Check file extension
        if not filename.lower().endswith('.pdf'):
            errors.append('File must have .pdf extension')
        
        return {
            'is_valid': len(errors) == 0,
            'errors': errors
        }

    def extract_pdf_metadata(self, pdf_content: bytes) -> Optional[dict]:
        """Extract metadata from PDF"""
        try:
            pdf_stream = BytesIO(pdf_content)
            pdf_reader = PyPDF2.PdfReader(pdf_stream)
            
            metadata = pdf_reader.metadata
            
            if metadata:
                return {
                    'title': metadata.get('/Title', 'Unknown'),
                    'author': metadata.get('/Author', 'Unknown'),
                    'subject': metadata.get('/Subject', 'Unknown'),
                    'creator': metadata.get('/Creator', 'Unknown'),
                    'producer': metadata.get('/Producer', 'Unknown'),
                    'creation_date': str(metadata.get('/CreationDate', 'Unknown')),
                    'modification_date': str(metadata.get('/ModDate', 'Unknown')),
                    'page_count': len(pdf_reader.pages),
                    'file_size': len(pdf_content)
                }
            else:
                return {
                    'title': 'No metadata available',
                    'author': 'Unknown',
                    'subject': 'Unknown',
                    'creator': 'Unknown',
                    'producer': 'Unknown',
                    'creation_date': 'Unknown',
                    'modification_date': 'Unknown',
                    'page_count': len(pdf_reader.pages),
                    'file_size': len(pdf_content)
                }
                
        except Exception as e:
            logger.error(f"Error extracting PDF metadata: {str(e)}")
            return None

    def is_pdf_password_protected(self, pdf_content: bytes) -> bool:
        """Check if PDF is password protected"""
        try:
            pdf_stream = BytesIO(pdf_content)
            pdf_reader = PyPDF2.PdfReader(pdf_stream)
            return pdf_reader.is_encrypted
        except Exception:
            return False

    def estimate_pdf_complexity(self, pdf_content: bytes) -> str:
        """Estimate PDF complexity for processing time"""
        try:
            size_in_mb = len(pdf_content) / (1024 * 1024)
            
            if size_in_mb < 1:
                return 'low'
            elif size_in_mb < 5:
                return 'medium'
            else:
                return 'high'
                
        except Exception:
            return 'unknown'